# Algorithm practice with LDA

**Note:**
- In this experiment, we will perform several operations on a CSV file using the NumPy, Pandas, Matplotlib, Seaborn, and SKLearn libraries.
- We are going to perform the LDA technique on the Dataset.
- Our primary focus will be on key operations commonly used in these libraries for building Logistic Regression models.
- The following operations will be conducted on the CSV file:

   - Importing the necessary libraries.
   - Loading the dataset for the experiment.
   - Understanding the data contained in the file.
   - Performing data cleaning and visualisation operations.
   - Preparing the data for further processing within the Data Science Life Cycle.
   - Conducting Exploratory Data Analysis (EDA) on the processed data.
   - Building predictive models using the EDA-processed data.
   - Performing model evaluations.

**Experiment Link** --> https://colab.research.google.com/drive/1tm6EMIV9h8e9-r-1-4rRpmG2dvcLH4lP?usp=sharing

### STEP 1. Problem statement:-
- We want to perform the dimensionality reduction LDA operation on the dataset to create new features by merging the existing features to train the supervised model.

### STEP 2.Importing the necessary libraries.

In [1]:
# Importing the required libraries.
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

### STEP 3.Loading the Dataset into the Environment.

In [2]:
from sklearn.datasets import load_digits

In [3]:
# Loading the dataset.
dataset = load_digits()

In [4]:
df = pd.DataFrame(dataset.data, columns=dataset.feature_names)

In [5]:
# Understanding the dataset.
df.head()

,pixel_0_0,pixel_0_1,pixel_0_2,pixel_0_3,pixel_0_4,pixel_0_5,pixel_0_6,pixel_0_7,pixel_1_0,pixel_1_1,...,pixel_6_6,pixel_6_7,pixel_7_0,pixel_7_1,pixel_7_2,pixel_7_3,pixel_7_4,pixel_7_5,pixel_7_6,pixel_7_7
0,0.0,0.0,5.0,13.0,9.0,1.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,6.0,13.0,10.0,0.0,0.0,0.0
1,0.0,0.0,0.0,12.0,13.0,5.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,11.0,16.0,10.0,0.0,0.0
2,0.0,0.0,0.0,4.0,15.0,12.0,0.0,0.0,0.0,0.0,...,5.0,0.0,0.0,0.0,0.0,3.0,11.0,16.0,9.0,0.0
3,0.0,0.0,7.0,15.0,13.0,1.0,0.0,0.0,0.0,8.0,...,9.0,0.0,0.0,0.0,7.0,13.0,13.0,9.0,0.0,0.0
4,0.0,0.0,0.0,1.0,11.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,0.0,2.0,16.0,4.0,0.0,0.0


In [6]:
df['Target'] = dataset.target

In [7]:
df.head()

,pixel_0_0,pixel_0_1,pixel_0_2,pixel_0_3,pixel_0_4,pixel_0_5,pixel_0_6,pixel_0_7,pixel_1_0,pixel_1_1,...,pixel_6_7,pixel_7_0,pixel_7_1,pixel_7_2,pixel_7_3,pixel_7_4,pixel_7_5,pixel_7_6,pixel_7_7,Target
0,0.0,0.0,5.0,13.0,9.0,1.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,6.0,13.0,10.0,0.0,0.0,0.0,0
1,0.0,0.0,0.0,12.0,13.0,5.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,11.0,16.0,10.0,0.0,0.0,1
2,0.0,0.0,0.0,4.0,15.0,12.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,3.0,11.0,16.0,9.0,0.0,2
3,0.0,0.0,7.0,15.0,13.0,1.0,0.0,0.0,0.0,8.0,...,0.0,0.0,0.0,7.0,13.0,13.0,9.0,0.0,0.0,3
4,0.0,0.0,0.0,1.0,11.0,0.0,0.0,0.0,0.0,0.0,...,0.0,0.0,0.0,0.0,2.0,16.0,4.0,0.0,0.0,4


In [8]:
df.shape

(1797, 65)

### STEP 4. Data Cleaning Operation performed on the dataset.


In [9]:
# checking for null values.
df.isnull().sum().sum()

np.int64(0)

In [10]:
# Checking for Duplicate Values.
df.duplicated().sum()

np.int64(0)

### STEP 5. Divide the dataset into features (x) and the target variable (y) for training the model.

In [11]:
x = df.drop('Target',axis=1)
y = df['Target']

In [12]:
from sklearn.model_selection import train_test_split

In [13]:
x_train , x_test , y_train,y_test = train_test_split(x,y,test_size=0.20,random_state=3)

In [14]:
x_train.shape

(1437, 64)

In [15]:
y_train.shape

(1437,)

In [16]:
x_test.shape

(360, 64)

### STEP 6.Feature Engineering
- For Encoding,Scaling,Feature creation.


**LDA -** Linear DiscriminantAnalysis

- It is a supervised machine learning technique.
- Its main goal is to maximise class separability.
- LDA works by finding a linear combination of features that best separates the classes.

**Process** -
- Apply standardisation on the x dataframe.
- selecting the n components for creating the LDA.


- To find the number of components required for LDA, use the formula.
- n_components = min(no.of features,no.of classes-1).
- EX:-

    - number of features = 64
    - number of classes = 10
    - min(64,10-1) => min(64,9)
    - 9 is the minimum, so we will select 9 components.

In [17]:
# STEP 1- Standardize the data
from sklearn.preprocessing import StandardScaler

In [18]:
sc = StandardScaler()
x_train = sc.fit_transform(x_train)
x_test = sc.transform(x_test)

- We do not use fit on x_test because we don't want  do not use fit on x_test because we don't to recalculate the mean and sd

In [19]:
x_train.std()


np.float64(0.9762812094883319)

In [20]:
x_test.std()

np.float64(0.9489205159243681)

In [21]:
pd.DataFrame(x_train)

,0,1,2,3,4,5,6,7,8,9,...,54,55,56,57,58,59,60,61,62,63
0,0.0,-0.332769,-0.241767,0.977815,0.976815,0.398811,-0.407540,-0.121535,-0.061417,-0.616625,...,-0.766521,-0.220588,-0.026389,-0.303466,0.098887,-0.026314,0.014679,-0.836130,-0.517331,-0.200105
1,0.0,-0.332769,-1.080439,-1.868958,0.506806,1.461074,-0.101885,-0.121535,-0.061417,-0.616625,...,-0.157963,-0.220588,-0.026389,-0.303466,-1.075498,-2.317994,0.841860,1.193463,-0.517331,-0.200105
2,0.0,-0.332769,-1.080439,0.028890,0.271801,-1.017540,-0.407540,-0.121535,-0.061417,-0.616625,...,-0.766521,-0.220588,-0.026389,-0.303466,-1.075498,-0.255482,0.221474,-1.174395,-0.517331,-0.200105
3,0.0,2.949996,1.435576,0.977815,-0.668218,-1.017540,-0.407540,-0.121535,-0.061417,3.455195,...,0.450596,0.850309,-0.026389,1.871500,1.860465,0.890358,0.841860,1.531728,3.333262,1.374078
4,0.0,-0.332769,1.016240,0.977815,-0.198209,-1.017540,-0.407540,-0.121535,-0.061417,1.889110,...,0.247743,-0.220588,-0.026389,-0.303466,1.273272,0.890358,0.841860,1.531728,3.092600,-0.200105
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1432,0.0,-0.332769,-0.661103,0.503353,-0.433213,-1.017540,-0.407540,-0.121535,-0.061417,-0.616625,...,2.276272,-0.220588,-0.026389,-0.303466,-0.684037,-0.255482,0.841860,1.362595,0.685979,-0.200105
1433,0.0,-0.332769,1.016240,-0.682803,-2.783261,-1.017540,-0.407540,-0.121535,-0.061417,-0.616625,...,0.450596,4.063002,-0.026389,-0.303466,0.294618,0.890358,0.841860,1.531728,3.333262,7.146081
1434,0.0,-0.332769,0.596904,0.028890,0.036796,-0.840496,-0.407540,-0.121535,-0.061417,0.323025,...,0.450596,-0.220588,-0.026389,-0.303466,-0.096844,0.202854,-0.192117,-0.497865,-0.517331,-0.200105
1435,0.0,1.855741,1.645244,0.977815,0.976815,0.752899,-0.407540,-0.121535,19.550963,4.081629,...,0.856302,0.850309,-0.026389,1.871500,1.860465,0.890358,0.841860,0.855197,2.370614,0.324623


In [22]:
 # Apply LDA
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis

In [23]:
LDA = LinearDiscriminantAnalysis(n_components=9)
x_train = LDA.fit_transform(x_train,y_train)
x_test = LDA.transform(x_test)

In [24]:
x_train.shape

(1437, 9)

In [25]:
x_test.shape

(360, 9)

### STEP 7.Model Selection & Training: Choose algorithms based on problem type.


In [26]:
# Train the model.
from sklearn.linear_model import LogisticRegression

In [27]:
model = LogisticRegression()
model.fit(x_train,y_train)

LogisticRegression()

In [28]:
# Predicting the values with test data.
model_pred = model.predict(x_test)

### STEP 8: Model Evaluation  
- Assessing the performance of the model based on various parameters.


In [29]:
from sklearn.metrics import *

In [30]:
accuracy_score(y_test,model_pred)

0.9444444444444444

- After building the ML model using the LDA data, we evaluated it and achieved " 94 % accuracy ".
